### ISSUE TITLE
Request header timeout after only 1 second

### ISSUE THREAD

**Issue by halter73** [aspnet/KestrelHttpServer#2850]
Receiving the request start line and headers and start line should only fail if it takes over 30 seconds by default, however [one of the Windows Jenkins validation runs](https://ci3.dot.net/job/aspnet_KestrelHttpServer/job/release_2.2/job/windows-Configuration_Debug_prtest/160/) for #2838 failed with a request header timeout after only one second.

```
16:37:03   Starting test group: Sockets.FunctionalTests/net461
16:37:03   Microsoft (R) Test Execution Command Line Tool Version 15.7.0
16:37:03   Copyright (c) Microsoft Corporation.  All rights reserved.
16:37:03   
16:37:03   Starting test execution, please wait...
16:37:06   [xUnit.net 00:00:02.44]     LargeUpload(contentLength: 2147483648, checkBytes: False) [FAIL]
16:37:06   Failed   LargeUpload(contentLength: 2147483648, checkBytes: False)
16:37:06   Error Message:
16:37:06    System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host
16:37:06   Stack Trace:
16:37:06      at System.Net.Sockets.Socket.Send(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
16:37:06      at System.Net.Sockets.Socket.Send(Byte[] buffer)
16:37:06      at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.RequestTests.LargeUpload(Int64 contentLength, Boolean checkBytes)
16:37:06   Standard Output Messages:
16:37:06    | [0.012s] TestLifetime Information: Starting test LargeUpload-2147483648-False at 2018-08-23T23:37:04
16:37:06    | [0.109s] Microsoft.AspNetCore.Hosting.Internal.WebHost Debug: Hosting starting
16:37:06    | [0.195s] Microsoft.AspNetCore.Hosting.Internal.WebHost Debug: Hosting started
16:37:06    | [0.196s] Microsoft.AspNetCore.Hosting.Internal.WebHost Debug: Loaded hosting startup assembly Sockets.FunctionalTests
16:37:06    | [0.301s] Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLG9A5APIEA1" started.
16:37:06    | [1.410s] Microsoft.AspNetCore.Server.Kestrel Information: Connection id "0HLG9A5APIEA1" bad request data: "Reading the request headers timed out."
16:37:06    | Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Reading the request headers timed out.
16:37:06    |    at Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException.Throw(RequestRejectionReason reason)
16:37:06    |    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.TryParseRequest(ReadResult result, Boolean& endConnection)
16:37:06    |    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.<ProcessRequests>d__189`1.MoveNext()
16:37:06    | --- End of stack trace from previous location where exception was thrown ---
16:37:06    |    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
16:37:06    |    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
16:37:06    |    at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
16:37:06    |    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.<ProcessRequestsAsync>d__188`1.MoveNext()
16:37:06    | [1.422s] Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLG9A5APIEA1" disconnecting.
16:37:06    | [1.423s] Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets Debug: Connection id "0HLG9A5APIEA1" sending FIN.
16:37:06    | [1.424s] Microsoft.AspNetCore.Hosting.Internal.WebHost Debug: Hosting shutdown
16:37:06    | [1.425s] Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLG9A5APIEA1" stopped.
16:37:06    | [1.428s] TestLifetime Information: Finished test LargeUpload-2147483648-False in 1.4230076s
16:37:06   
16:37:06   
16:38:03   Test Run Failed.
16:38:03   Results File: D:\j\workspace\windows-Confi---058f2efb\artifacts\logs\Sockets.FunctionalTests-net461-160.trx
16:38:03   
16:38:03   Total tests: 67. Passed: 66. Failed: 1. Skipped: 0.
16:38:03   Test execution time: 51.6766 Seconds
```

This appears to be caused by a race introduced by #2834 where the first calls to `HeartbeatManager.UtcNow` returns its default value of [1/1/0001 12:00:00 AM +00:00] causing all timeouts that have been configured prior to the first tick to the fire when the first tick fires.

The reason this wasn't caught by our unit/inmemory tests was because TestServiceContext.SystemClock was set to a SystemClock instance instead of an HeartbeatManager instance like it is in KestrelServer.

**Comment by davidfowl** [comment]
@dougbu why did you re-open this?

**Comment by dougbu** [comment]
@davidfowl you type faster than me 😄 I was just typing more information in.

Because it's still happening with Kestrel in at least one flaky test. See last failure in aspnet/MusicStore#848, the details of that failure (http://aspnetci/viewLog.html?buildId=552876&buildTypeId=XPlat_Mac_MacUniverseCoreCLR), and the low (95.5%) reliability of that test (http://aspnetci/project.html?projectId=XPlat&tab=testDetails&testNameId=-322587869374600688#analysis).

Basically, I'm wondering if the fix here was incomplete.
